src/main/java/mcscheduler/logic/commands/AssignCommand.java

```diff
@@ -91,12 +96,27 @@ public CommandResult execute(Model model) throws CommandException {
                         Messages.MESSAGE_INVALID_ASSIGNMENT_UNAVAILABLE, workerToAssign.getName(), shiftToAssign));
             }
 
+            /*
             if (!shiftToAssign.isRoleRequired(role)) {
                 throw new CommandException(Messages.MESSAGE_INVALID_ASSIGNMENT_NOT_REQUIRED);
             }
+             */
+            // Count for RoleRequirements
+            if (!(role instanceof Leave)) {
+                if (!requiredRoles.containsKey(role)) {
+                    requiredRoles.put(role, getQuantityRequiredForRole(shiftToAssign, role));
+                }
+                requiredRoles.put(role, requiredRoles.get(role) - 1);
+                if (requiredRoles.get(role) < 0) {
+                    // This assignCommand will exceed the role's requirement
+                    throw new CommandException(
+                        String.format(Messages.MESSAGE_INVALID_ASSIGNMENT_NOT_REQUIRED, role, shiftToAssign));
+                }
+            }
 
             Assignment assignmentToAdd = new Assignment(shiftToAssign, workerToAssign, role);
-            if (model.hasAssignment(assignmentToAdd)) {
+            // Prevent duplicates in model & Prevent duplicates in a single call
+            if (model.hasAssignment(assignmentToAdd) || assignmentsToAdd.contains(assignmentToAdd)) {
```

Comment:
Same comment as above, I think we can just place this check in the `for` loop before adding each assignment as well.